### PR TITLE
[#162000] SciShield local storage

### PR DESCRIPTION
**Pull request by jossim**
# Release Notes

Some instruments require training before users are allowed to use them. The SciShield API can be used to verify a user has the proper training to use the instrument. But, if the API is down, the user will get a 500 error instead of being able to reserve the instrument.

This PR creates a local copy of a user's trainings that are checked before the API, allowing the system to be used even if the API is down.

### CODE REVIEW

File: app/services/research_safety_adapters/scishield_training_synchronizer.rb [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+      @users = User.select(:id, :email)
+    end
+
+    def synchronize
+      return if api_unavailable?
+
+      ScishieldTraining.transaction do
+        ScishieldTraining.delete_all
+
+        @users.each do |user|
```
Comment (jossim) [inline app/services/research_safety_adapters/scishield_training_synchronizer.rb:16]:
```suggestion
        @users.find_each do |user|
```

File: spec/services/scishield_api_adapter_spec.rb [hunk 2]
```diff
@@ -83,4 +83,17 @@
     end
   end
 
+  describe "Scishield training exists in database" do
```
Comment (jossim) [inline spec/services/scishield_api_adapter_spec.rb:86]: Change this to `context`, and make a context for empty db

File: app/services/research_safety_adapters/scishield_api_client.rb [hunk 3]
```diff
@@ -35,12 +35,15 @@ def keys_present?
     end
 
     def certifications_for(email)
+      training_api_request(email).body
```
Comment (jossim) [inline app/services/research_safety_adapters/scishield_api_client.rb:38]:
```suggestion
      response = training_api_request(email)
      response.body
```

File: app/services/research_safety_adapters/scishield_training_synchronizer.rb [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+          adapter = ScishieldApiAdapter.new(user)
+          cert_names = adapter.certified_course_names_from_api
+
+          cert_names.each { |cert_name| ScishieldTraining.create(user_id: user.id, course_name: cert_name) }
+        end
+      rescue StandardError
```
Comment (jossim) [inline app/services/research_safety_adapters/scishield_training_synchronizer.rb:22]: Put in logging and Rollbar here

File: db/migrate/20231116155708_create_scishield_trainings.rb [hunk 5]
```diff
@@ -0,0 +1,10 @@
+class CreateScishieldTrainings < ActiveRecord::Migration[7.0]
```
Comment (jossim) [inline db/migrate/20231116155708_create_scishield_trainings.rb:1]:
```suggestion
# frozen_string_literal: true

class CreateScishieldTrainings < ActiveRecord::Migration[7.0]
```

File: db/migrate/20231116155708_create_scishield_trainings.rb [hunk 6]
```diff
@@ -0,0 +1,10 @@
...
+  def change
+    create_table :scishield_trainings do |t|
+      t.integer :user_id
```
Comment (jossim) [inline db/migrate/20231116155708_create_scishield_trainings.rb:4]: add null false

File: spec/factories/scishield_trainings.rb [hunk 7]
```diff
@@ -0,0 +1,8 @@
+# frozen_string_literal: true
+
+FactoryBot.define do
+  factory :scishield_training do
+    user_id { 1 }
```
Comment (jossim) [inline spec/factories/scishield_trainings.rb:5]:
```suggestion
    user_id { User.first.id }
```

File: spec/services/research_safety_adapters/scishield_training_synchronizer_spec.rb [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+
+    it "adds courses to database" do
+      expect(ScishieldTraining.count).to eq 0
+      synchronizer.synchronize
+      expect(ScishieldTraining.count).to eq 3
+      expect(ScishieldTraining.all.map(&:course_name)).to include(*course_names)
+    end
+  end
+
+  context "when the API reponses with error" do
```
Comment (jossim) [inline spec/services/research_safety_adapters/scishield_training_synchronizer_spec.rb:35]: Add some courses here, and test that they aren't removed

**Comment by jossim** [comment]
This still needs a cron job setup

File: app/models/scishield_training.rb [hunk 9]
```diff
@@ -0,0 +1,5 @@
+# frozen_string_literal: true
+
+class ScishieldTraining < ApplicationRecord
```
Comment (giladshanan) [inline app/models/scishield_training.rb:3]: Please add a code comment explaining the intent/responsibility of this new class

File: app/services/research_safety_adapters/scishield_training_synchronizer.rb [hunk 10]
```diff
@@ -0,0 +1,45 @@
+# frozen_string_literal: true
+
+module ResearchSafetyAdapters
+
+  class ScishieldTrainingSynchronizer
```
Comment (giladshanan) [inline app/services/research_safety_adapters/scishield_training_synchronizer.rb:5]: Please add a code comment explaining the intent of this class

File: app/services/research_safety_adapters/scishield_training_synchronizer.rb [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+    def initialize
+      @users = User.active.select(:id, :email)
```
Comment (giladshanan) [inline app/services/research_safety_adapters/scishield_training_synchronizer.rb:7]: Why are you just grabbing id and email here?  Is there a performance concern?  

We might want to know the username at some point.  Also, `ScishieldApiAdapter` depends on this data including everything it needs to know about users.  Right now that's not encoded anywhere.  I'd vote for either just grabbing all the user data like `User.active` or defining the list of attributes in `ScishieldApiAdapter` and the using it here like:
`@users = User.active.select(ScishieldApiAdapter.user_attributes)`.

File: app/services/research_safety_adapters/scishield_training_synchronizer.rb [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+    end
+
+    def synchronize
+      return if api_unavailable?
```
Comment (giladshanan) [inline app/services/research_safety_adapters/scishield_training_synchronizer.rb:11]: I think we will want to know if the API is down - can you add a `Rollbar.error` here?

File: spec/services/research_safety_adapters/scishield_training_synchronizer_spec.rb [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+
+  before do
+    stub_request(:get, api_endpoint)
+      .to_return(
+        status: status_code,
+        body: File.new(response)
+      )
+  end
+
+  context "when the API responses without error" do
```
Comment (giladshanan) [inline spec/services/research_safety_adapters/scishield_training_synchronizer_spec.rb:20]:
```suggestion
  context "when the API responds without error" do
```

File: spec/services/research_safety_adapters/scishield_training_synchronizer_spec.rb [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+    let(:status_code) { "200" }
+
+    it "adds courses to database" do
+      expect(ScishieldTraining.count).to eq 0
+      synchronizer.synchronize
+      expect(ScishieldTraining.count).to eq 3
+      expect(ScishieldTraining.all.map(&:course_name)).to include(*course_names)
```
Comment (giladshanan) [inline spec/services/research_safety_adapters/scishield_training_synchronizer_spec.rb:27]: Have you tried the [`contain_exactly`](https://medium.com/beambenefits/array-matchers-in-rspec-42b370e3983e) matcher here?
```suggestion
      expect(ScishieldTraining.all.map(&:course_name)).to contain_exactly(*course_names)
```

File: spec/services/research_safety_adapters/scishield_training_synchronizer_spec.rb [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+    end
+  end
+
+  context "when the API reponses with error" do
```
Comment (giladshanan) [inline spec/services/research_safety_adapters/scishield_training_synchronizer_spec.rb:31]:
```suggestion
  context "when the API responds with error" do
```

File: spec/services/research_safety_adapters/scishield_training_synchronizer_spec.rb [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+    let(:status_code) { 500 }
```
Comment (giladshanan) [inline spec/services/research_safety_adapters/scishield_training_synchronizer_spec.rb:32]: Can you add specs for 404 and 403?

File: spec/services/scishield_api_adapter_spec.rb [hunk 17]
```diff
@@ -4,83 +4,98 @@
...
-          status: 403,
-        )
+      3.times { create(:scishield_training, user_id: user.id) }
     end
 
-    it "raises an error" do
-      expect { adapter.certified?(anything) }.to raise_error(/Forbidden: This route can only be accessed by authenticated users/)
+    it "gets course names from database" do
+      expect(adapter.certified_course_names).to contain_exactly(
+        "Scishield Course Name 1",
```
Comment (giladshanan) [inline spec/services/scishield_api_adapter_spec.rb:95]: These course names depend on the factory - which can be a little brittle over time.   I'd vote for either setting the names explicitly in the spec, or using `let!` to create the trainings and then referencing that variable to get the expected names.

**Review by giladshanan — CHANGES_REQUESTED** [review]
Left you some suggestions, looks good!

File: app/services/research_safety_adapters/scishield_training_synchronizer.rb (hunk 11 above)
Comment (jossim) [reply app/services/research_safety_adapters/scishield_training_synchronizer.rb:10]: Yeah, I was thinking about performance

**Review by giladshanan — APPROVED** [review]
Looks good, just a couple non-blocking comments

**Pull request merged**